HW-0000 Travel agency service

Kiraberos opened the pull request:
- created a travel booking solution

Kiraberos committed:
HW-0000 Travel agency service

- created a travel booking solution

Puchkof commented on TravelAgencyService/Inerfaces/ITravelAgency.cs:
```diff
@@ -0,0 +1,9 @@
+namespace TravelAgencyService;
+
+public class ITravelAgency
```
Did you want to create interface here? Or class?

Puchkof replied:
If interface - it should be interface, instead I would make it abstract to avoid creation of such instances

Puchkof commented on TravelAgencyService/Models/AgencyModel.cs:
```diff
@@ -0,0 +1,63 @@
+
+    public AgencyModel()
+    {
+        _destinations = new List<DestinationProcessingService.Destination>
```
You can set this values on field level, without constructor. It will work absolutely the same

Puchkof commented on TravelAgencyService/Models/AgencyModel.cs:
```diff
@@ -0,0 +1,63 @@
+
+        if (destination == null)
+        {
+            throw new Exception("Wrong destination. Please, choose the right one.");
```
Try to create custom exception like `InvalidDestinationException`

Puchkof commented on TravelAgencyService/Models/AgencyModel.cs:
```diff
@@ -0,0 +1,63 @@
+
+        if (!bookings.Any())
+        {
+            Console.WriteLine("There are no current bookings");
```
I would avoid coupling with presentation layer in domain layer. You can try just to return all bookings here and display somewhere else

Puchkof commented on TravelAgencyService/Models/Execute.cs:
```diff
@@ -0,0 +1,40 @@
+﻿namespace TravelAgencyService;
+
+class Execute: ITravelAgency
```
I would call it just `TravelAgency`

Puchkof commented on TravelAgencyService/Models/Execute.cs:
```diff
@@ -0,0 +1,40 @@
+            switch (choice)
+            {
+                case "1":
+                    new AgencyOperationsService().BookTicket(travelAgency);
```
Maybe you can create instance of `AgencyOperationsService` once before switch?

Puchkof commented on TravelAgencyService/Services/DestinationProcessingService.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            if (CurrentBookings >= MaxBookings)
+            {
+                throw new Exception("There are no available reservations for this trip. Please, try again later.");
```
Also try custom exception here, its better practice on commercial projects

Kiraberos committed:
HW-0000 Travel agency service

- refactored the current solution according to remarks

Puchkof commented on TravelAgencyService/Models/AgencyModel.cs:
```diff
@@ -0,0 +1,50 @@
+
+            if (destination == null)
+            {
+                throw new ExceptionMgmt.InvalidDestinationException("Wrong destination. Please, choose the right one.");
```
Its not a very convenient way to use Exceptions in C#, just throw with new and thats enough

Puchkof commented on TravelAgencyService/Services/Destination.cs:
```diff
@@ -0,0 +1,27 @@
+    public class Destination(string name, int maxBookings)
+    {
+        public string Name { get; } = name;
+        private int MaxBookings { get; } = maxBookings;
```
I think it should be read only field _maxBookings, not property

Puchkof commented on TravelAgencyService/Models/TravelAgency.cs:
```diff
@@ -0,0 +1,45 @@
+            _displaySelectorService = new ActionSelectorService();
+            _agencyOperation = new AgencyOperationsService();
+            _displaySelectorService.DisplayMenu();
+            string choice = Console.ReadLine();
```
Maybe its better to create method GetChoice in displaySelectorService ?

Puchkof commented on TravelAgencyService/Services/AgencyOperationsService.cs:
```diff
@@ -0,0 +1,52 @@
+       
+    public void BookTicket(AgencyModel travelAgency)
+    {
+        Console.Write("Enter the name of the destination: ");
```
I would move all the console operations to separate class and leave only business logic here

Puchkof commented on TravelAgencyService/Services/AgencyOperationsService.cs:
```diff
@@ -0,0 +1,52 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.Message);
```
Should we do some retry logic here?

Kiraberos committed:
HW-0000 Travel agency service

- created sln file

Kiraberos committed:
HW-0000 Travel agency service

- configure sln file

Kiraberos committed:
HW-0000 Travel agency service

-  moved all console operations to the separate console service, also made _maxBooking field read-only

Kiraberos committed:
HW-0000 Travel agency service

- expanded the console service left only business logic

Kiraberos committed:
HW-0000 Travel agency service

- implemented retry logic is in case we catch an exception

Puchkof commented on TravelAgencyService/TravelAgencyService/Models/TravelAgency.cs:
```diff
@@ -0,0 +1,43 @@
+                    exit = true;
+                    break;
+                default:
+                    Console.WriteLine("Wrong choice. Please try again");
```
Maybe you need to use consoleService here?

Puchkof commented on TravelAgencyService/TravelAgencyService/Services/AgencyOperationsService.cs:
```diff
@@ -0,0 +1,57 @@
+                retryAttempts++;
+                if (retryAttempts < maxRetryAttempts)
+                {
+                    Thread.Sleep(delayMilliseconds);
```
You can also use async analogue Task.Delay but lets keep as it is as you haven't learn async/await yet
